src/main/java/filter/expression/Qualifier.java

```diff
@@ -717,6 +714,61 @@ private boolean milestoneSatisfies(IModel model, TurboIssue issue) {
         String contents = content.get().toLowerCase();
         String title = milestone.get().getTitle().toLowerCase();
 
+        List<TurboMilestone> allMilestones = model.getMilestones();
+        Collections.sort(allMilestones, new TurboMilestoneDueDateComparator());
+
+        int currentIndex = -1;
+
+        for (TurboMilestone checker : allMilestones) {
+            // current milestone must be open
+            if (checker.isOpen()) {
+                // if milestone is overdue but have no open issue, ignore
+                if (checker.getDueDate().isPresent() &&
+                        checker.getDueDate().get().isBefore(LocalDate.now()) &&
+                        checker.getOpenIssues() == 0) {
+                    continue;
+                }
+                currentIndex = allMilestones.indexOf(checker);
+                break;
+            }
+        }
+
+        // if no open milestone, set current as last milestone
+        if (currentIndex == -1 && allMilestones.size() > 0) {
+            currentIndex = allMilestones.size() - 1;
+        }
+
+        String alias = contents;
+
+        if (alias.equalsIgnoreCase("current") ||
+                alias.equalsIgnoreCase("curr")) {
+            contents = allMilestones.get(currentIndex).getTitle();
+        } else if (alias.equalsIgnoreCase("previous") ||
+                alias.equalsIgnoreCase("prev")) {
+            if (currentIndex - 1 >= 0) {
+                contents = allMilestones.get(currentIndex - 1).getTitle();
+            }
+        } else if (alias.equalsIgnoreCase("next")) {
+            if (currentIndex + 1 < allMilestones.size()) {
+                contents = allMilestones.get(currentIndex + 1).getTitle();
+            }
+        } else if (alias.startsWith("previous-") ||
+                alias.startsWith("prev-")) {
+            int offsetPos = alias.indexOf('-') + 1;
+            int offset = Integer.parseInt(alias.substring(offsetPos).trim());
+            if (currentIndex - 1 - offset >= 0) {
+                contents = allMilestones.get(currentIndex - 1 - offset).getTitle();
+            }
+        } else if (alias.startsWith("next+")) {
+            int offsetPos = alias.indexOf('+') + 1;
+            int offset = Integer.parseInt(alias.substring(offsetPos).trim());
+            if (currentIndex + 1 + offset < allMilestones.size()) {
+                contents = allMilestones.get(currentIndex + 1 + offset).getTitle();
+            }
+        }
+
+        contents = contents.toLowerCase();
+
```

Comment:
Okay, will do the regex. I will changed everyhig to just current/curr +/-
so no need for prev or next.

Just to confirm, so should I do the sorting and getting all milestones of a
repo/repos outside Qualifier?

About cleaning up helper methods, which ones are you talking about? Like
the `milestoneAliasConverter`?
On Oct 13, 2015 8:44 AM, "Darius Foo" <notifications@github.com> wrote:

> In src/main/java/filter/expression/Qualifier.java
> <https://github.com/HubTurbo/HubTurbo/pull/1050#discussion_r41816857>:
>
> > +                alias.startsWith("prev-")) {
> > +            int offsetPos = alias.indexOf('-') + 1;
> > +            int offset = Integer.parseInt(alias.substring(offsetPos).trim());
> > +            if (currentIndex - 1 - offset >= 0) {
> > +                contents = allMilestones.get(currentIndex - 1 - offset).getTitle();
> > +            }
> > +        } else if (alias.startsWith("next+")) {
> > +            int offsetPos = alias.indexOf('+') + 1;
> > +            int offset = Integer.parseInt(alias.substring(offsetPos).trim());
> > +            if (currentIndex + 1 + offset < allMilestones.size()) {
> > +                contents = allMilestones.get(currentIndex + 1 + offset).getTitle();
> > +            }
> > +        }
> > +
> > +        contents = contents.toLowerCase();
> > +
>
> How about something like (curr(?:ent)?|prev(?:ious)?)(?:(\+|-)1)?? \1
> will tell you the first word, and \2 will tell you whether the offset is
> positive or negative, or zero if it's not present.
>
> I guess you would need another method to be used alongside
> Qualifier::process. Some of it should also be moved out there. The new
> method could return a MetaQualifierInfo and/or a FilterExpression... or
> maybe we need two methods.
>
> Find contents of repo qualifiers, then use that to filter the list of
> milestones? If we have something like repo:a || repo:b we will have to
> get milestones across repos.
>
> On a side note we should clean up the helper methods in Qualifier... Make
> them higher-level and better at doing things like this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HubTurbo/HubTurbo/pull/1050/files#r41816857>.
>